### PR TITLE
Implement flow analysis support for tuples and deconstruction assignment

### LINKED ISSUE #1571: Pending operations to handle in all the dataflow analyses
- [x] Patterns
- [x] Tuples, deconstruction
- [X] Try/catch/finally
- [x] Lambda/local function invocations
- [ ] IStaticLocalInitializationSemaphoreOperation

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #1571 

Future work:
1. Predicate analysis for ITupleBinaryOperation
2. Deconstruction assignment from a non-tuple typed entity with a Deconstruct method. The IOperation API for deconstruction does not expose the Deconstruct method that is bound by the compiler.

### CODE REVIEW

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs [hunk 1]
```diff
@@ -113,55 +113,34 @@ private void HandleDisposingOperation(IOperation disposingOperation, IOperation
                 }
             }
 
-            private void HandlePossibleEscapingOperation(IOperation escapingOperation, IOperation escapedInstance)
+            private void HandlePossibleEscapingOperation(IOperation escapingOperation, ImmutableHashSet<AbstractLocation> escapedLocations)
             {
-                Debug.Assert(escapingOperation != null);
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs]: Majority of code tracking abstract locations/allocations that escaped the method body was moved from Dispose analysis to PointsTo analysis. This was required as we need to mark all the child tuple elements of escaped tuples as escaped, which needs PointsTo data. Additionally, this information about which creation locations escape a method seem to be generic enough to be useful outside dispose analysis, so it made sense to move it to the PointsToAnalysisResult (which is now a new type).

**Comment by mavasani** [comment]
Ping, I have another PR and another potential PR based on this, so would appreciate some reviews.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetAnalysis.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 using System;
 using System.Collections.Immutable;
+using System.Threading;
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetAnalysis.cs]: Nit: Why add this using?

Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetAnalysis.cs]: Hmm looks like I probably added something and reverted, but forgot to revert the using. Will remove, thanks.

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 3]
```diff
@@ -255,11 +294,30 @@ protected override void PostProcessArgument(IArgumentOperation operation, bool i
             {
                 base.PostProcessArgument(operation, isEscaped);
 
-                if (!isEscaped &&
-                    operation.Parameter.RefKind != RefKind.None &&
-                    AnalysisEntityFactory.TryCreate(operation, out var analysisEntity))
+                if (!isEscaped)
+                {
+                    // Update abstract value for unescaped ref or out argument (interprocedural analysis case).
+                    if (operation.Parameter.RefKind != RefKind.None &&
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Do you have to worry about RefKind.In here?

Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Good point, I should update the check to match the comment. Given that the abstract value for `in` arguments cannot change after the invocation, we should probably be fine either way.

**Comment by mavasani** [comment]
Going to address Paul's concern with a followup commit in https://github.com/dotnet/roslyn-analyzers/pull/2146

**Pull request merged**